Python/marshal.c

```diff
@@ -895,51 +991,43 @@ r_PyLong(RFILE *p)
         return NULL;
     }
 
-    size = 1 + (Py_ABS(n) - 1) / PyLong_MARSHAL_RATIO;
-    shorts_in_top_digit = 1 + (Py_ABS(n) - 1) % PyLong_MARSHAL_RATIO;
-    ob = _PyLong_New(size);
-    if (ob == NULL)
-        return NULL;
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+    Py_ssize_t marshal_ratio = layout->bits_per_digit/PyLong_MARSHAL_SHIFT;
 
-    _PyLong_SetSignAndDigitCount(ob, n < 0 ? -1 : 1, size);
+    /* must be a multiple of PyLong_MARSHAL_SHIFT */
+    assert(layout->bits_per_digit % PyLong_MARSHAL_SHIFT == 0);
 
-    for (i = 0; i < size-1; i++) {
-        d = 0;
-        for (j=0; j < PyLong_MARSHAL_RATIO; j++) {
-            md = r_short(p);
-            if (md < 0 || md > PyLong_MARSHAL_BASE)
-                goto bad_digit;
-            d += (digit)md << j*PyLong_MARSHAL_SHIFT;
-        }
-        ob->long_value.ob_digit[i] = d;
+    /* other assumptions on PyLongObject internals */
+    assert(layout->bits_per_digit <= 32);
```

Comment:
You might want to add the assertion 

```c
assert(layout->bits_per_digit >= PyLong_MARSHAL_SHIFT);
```

because otherwise `marshal_ratio` is 0 and you get a division by zero when computing `size` and a UB when taking `% 0`.